### PR TITLE
Bring offline KD upgrades such as Ghost Token and Top-P to Megatron K…

### PR DESCRIPTION
**Pull request by AAnoosheh**
### What does this PR do?

Type of change: New Feature

- Top-P feature in Megatron KD loss
- Ghost-token feature in Megatron KD loss
- More intuitive and better loss balancing scheme with `kd_loss_alpha` parameter

### Usage

```python
# Add a code snippet demonstrating how to use this
```

### Testing
Newly-expended unit tests

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: N/A
- Did you write any new necessary tests?: ✅ 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅ 
- Did you get Claude approval on this PR?: 

## Summary by CodeRabbit

* **New Features**
  * Added configurable Top-P (nucleus) filtering for Top-K KL distillation, including a minimum retained token count and optional ghost token.
  * Added `kd_loss_alpha` to control the convex combination of language-model and distillation losses.
  * Improved KL normalization across distributed model partitions and added temperature-scaled logits handling.

* **Breaking Changes**
  * Full-vocabulary KL normalization now uses a default ghost token.
  * Deprecated loss-scaling and skip-language-model options are ignored with warnings; use `kd_loss_alpha` instead.

### CODE REVIEW

File: examples/megatron_bridge/distill.py [hunk 1]
```diff
@@ -172,9 +173,23 @@ def get_args():
...
+        help="DEPRECATED and ignored. Whether the LM loss is skipped is derived from --kd_loss_alpha "
+        "(skipped iff alpha == 1.0).",
+    )
+    parser.add_argument(
+        "--kd_loss_alpha",
+        type=float,
+        default=0.9,
+        help="KD loss weight alpha in (1 - alpha) * lm_loss + alpha * kd_loss. 1.0 skips the LM loss entirely.",
+    )
+    parser.add_argument(
```
Comment (jenchen13) [inline examples/megatron_bridge/distill.py:187]: nit: move this arg before `--kd_loss_alpha` to be with the other deprecated arg

File: CHANGELOG.rst [hunk 2]
```diff
@@ -15,6 +15,7 @@ Changelog
 
 *Megatron Framework (M-LM / M-Bridge)*
 
+- Add optional Top-P (nucleus) truncation to the Megatron ``TopKLogitsKLLoss`` via ``logit_kl_top_p`` and ``logit_kl_top_p_min_k`` in ``DistillationConfig``: after the global Top-K selection, only the smallest prefix whose cumulative teacher probability reaches ``top_p`` (with a floor of ``min_k`` entries) contributes to the KL, mirroring ``--logits-save-top-p`` / ``--logits-save-top-p-min-k`` in Megatron-LM's logits saver.
```
Comment (jenchen13) [inline CHANGELOG.rst:18]: we should allow either top K _or_ top p sampling, not top K followed by top p <-- this is not true nucleus sampling and doesn't match inference time sampling

File: examples/megatron_bridge/distill.py [hunk 3]
```diff
@@ -435,9 +462,10 @@ def _build_model_provider(hf_path, load_weights=True, moe_grouped_gemm=True):
         )
 
     kd_config = ModelOptDistillConfig(
-        skip_lm_loss=not args.no_skip_lm_loss,
-        kd_loss_scale=args.kd_loss_scale,
+        kd_loss_alpha=args.kd_loss_alpha,
         logit_kl_topk=args.logit_kl_topk,
+        logit_kl_top_p=args.logit_kl_top_p,
+        logit_kl_top_p_min_k=args.logit_kl_top_p_min_k,
```
[the remark this answers is not in the document]
Comment (AAnoosheh) [reply examples/megatron_bridge/distill.py:480]: The two deprecated flags now emit a `FutureWarning` when passed. On the ghost-token flag: not adopted. The ghost token is now always on, and the `logit_kl_ghost_token` config field has been removed, so there is no opt-out left to expose.

File: modelopt/torch/distill/plugins/megatron.py [hunk 4]
```diff
@@ -445,14 +503,39 @@ def forward(self, predictions: Tensor, targets: Tensor) -> Tensor:
...
+
+        # Ghost token: residual probability mass outside the kept entries, for both distributions.
+        if self.add_ghost_token:
+            eps = 1e-8
+            student_kept_mass = (student_logp.exp() * mask).sum(dim=-1, keepdim=True)
+            teacher_kept_mass = (teacher_logp.exp() * mask).sum(dim=-1, keepdim=True)
+            student_residual = torch.log((1.0 - student_kept_mass).clamp(min=eps))
+            teacher_residual = torch.log((1.0 - teacher_kept_mass).clamp(min=eps))
+            student_logp = torch.cat([student_logp, student_residual], dim=-1)
+            teacher_logp = torch.cat([teacher_logp, teacher_residual], dim=-1)
```
[the remark this answers is not in the document]
Comment (AAnoosheh) [reply modelopt/torch/distill/plugins/megatron.py]: Adopted the `log(-expm1(log_kept))` form. The test references stay on the independent `log1p(-mass)` formula on purpose, since a reference that reuses the implementation expression would not catch a bug in it. With the accurate residual, the tolerances are tightened from 1e-5 to 1e-6 and pass.

File: modelopt/torch/distill/plugins/megatron.py [hunk 5]
```diff
@@ -76,8 +91,44 @@ def __post_init__(self):
...
+        if self.skip_lm_loss is not None:
+            if self.skip_lm_loss and self.kd_loss_alpha != 1.0:
+                warnings.warn(
+                    "DistillationConfig.skip_lm_loss is deprecated; translating skip_lm_loss=True to "
+                    "kd_loss_alpha=1.0. Set `kd_loss_alpha` directly instead.",
+                    FutureWarning,
+                    stacklevel=2,
+                )
+                self.kd_loss_alpha = 1.0
+            elif not self.skip_lm_loss and self.kd_loss_alpha == 1.0:
```
[the remark this answers is not in the document]
Comment (AAnoosheh) [reply modelopt/torch/distill/plugins/megatron.py]: Not adopting. Before this PR, `skip_lm_loss=True` meant the total loss was the KD loss alone, which is exactly `kd_loss_alpha=1.0` under the new convex combination. Translating it to 1.0 is what preserves the old behavior. Keeping the derived default of 0.9 would add the LM loss to runs that previously skipped it. The translation now applies only when `kd_loss_alpha` is not set explicitly, per the other review thread on this block.

File: examples/megatron_bridge/distill.py [hunk 6]
```diff
@@ -188,6 +203,19 @@ def get_args():
         help="Restrict the logit KL loss to the teacher's top-k vocabulary entries, "
         "replacing the full-vocab temporaries with [seq, k] ones.",
```
[the remark this answers is not in the document]
Comment (AAnoosheh) [reply examples/megatron_bridge/distill.py]: Fixed the `--logit_kl_topk` help text. It now describes the top-k plus residual-bucket restriction and notes that normalization is still over the full vocab. I left the class docstring note unchanged. It refers to the per-rank top-k all-gather used for tensor parallelism, which is still accurate.

File: CHANGELOG.rst (hunk 2 above)
Comment (AAnoosheh) [reply CHANGELOG.rst:26]: Keeping top-P applied within the top-K. When both are set, inference stacks combine them this way: HF generation and vLLM apply the top-K filter first, then the nucleus filter. It also mirrors the Megatron-LM offline logits saver (`--logits-save-top-k` then `--logits-save-top-p`), so live and cached KD targets match. A standalone nucleus would also need a global sort over the full vocab across TP shards, and `logit_kl_topk` keeps that bounded. Setting a large `logit_kl_topk` approximates pure top-P, since the nucleus rarely reaches that deep.
